### PR TITLE
Feat: Implement edit locked expenditure motion saga

### LINKED ISSUE #1976: Advanced Payments API: Changes to expenditures after locking (using motions)
Related specs from the main issue (#1956):
> As a USER, I would like to be able to make changes to the advanced payment during the process. However, I want it to require collective input after an initial review step.

> (Stages) Changes
> - Conditional stage that only appears if a change request has been made on the payment after the review stage.
> - Additional details are included in “Changes to advanced payments” below.

Changes to expenditures after locking can be made using `setExpenditureState` contracts function. This issue is to implement a `editLockedExpenditureMotion` saga that will accept an array of updated payouts and create a motion which, upon finalizing, will call the contracts function and update the expenditure accordingly.

Any motions created this way should be saved under `motions` field of `Expenditure` model.

The changes proposed via a motion should be stored and made available to the UI.

### PR DESCRIPTION
**Pull request by iamsamgibbs**
## Description

[Related block-ingestor PR](https://github.com/JoinColony/block-ingestor/pull/189)

Changes to expenditures after locking can be made using `setExpenditureState` contracts function. This PR implements an `editLockedExpenditureMotion` saga that accepts an array of updated payouts and creates a motion which, upon finalizing, calls the contracts function and updates the expenditure accordingly.

## Testing

1. Install the reputation weighting extension
2. Create a new expenditure (The expenditure ID will be logged to the console)
3. Enter the expenditure ID and lock the expenditure.
4. Run this query to verify the expenditure has been created with no motions attached, and one payout.

```
query MyQuery {
  listExpenditures(filter: {nativeId: {ge: 33}}) {
    items {
      nativeId
      motions {
        items {
          motionStateHistory {
            hasPassed
          }
          editedExpenditureSlots {
            claimDelay
            id
            payoutModifier
            recipientAddress
            payouts {
              amount
              isClaimed
              networkFee
              tokenAddress
            }
          }
        }
      }
      slots {
        payouts {
          amount
          isClaimed
          networkFee
          tokenAddress
        }
        claimDelay
        recipientAddress
        id
        payoutModifier
      }
      status
    }
  }
}

```

5. Click the 'Edit via motion' button and follow the link logged in the console
6. Run the query again to verify that the 'editedExpenditureSlots' field of the attached motion is correctly storing all of the proposed changes
7. Stake both sides and vote to support the motion. Then finalize the motion.
8. Run the query a third time to verify that the motion has passed, and the payouts have updated.

## Diffs

**New stuff** ✨

* EditLockedExpenditureMotion saga

Resolves #1976

### CODE REVIEW

**Review by ArmandoGraterol — APPROVED** [review]
Nice work Sam! Working as expected for me 👍

**Review by rdig — APPROVED** [review]
All good codewise, can't run this manually however.

I managed to get around the block ingestor by starting it manually from it's own repo, but then the amplify image won't build.

Screenshot from 2024-03-05 22-14-29

Seems like npm is either responding slowly or just down completely.

I'm actually surprised GH's UI still works.

At any rate, nice work here!

File: src/redux/sagas/motions/expenditures/editLockedExpenditureMotion.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    colonyAddress,
+  );
+
+  const batchId = 'motion-edit-locked-expenditure';
+
+  /**
+   * @NOTE: Resolving payouts means making sure that for every slot, there's only one payout with non-zero amount.
+   * This is to meet the UI requirement that there should be one payout per row.
+   */
+  const resolvedPayouts: ExpenditurePayoutFieldValue[] = [];
```
Comment (jakubcolony) [inline src/redux/sagas/motions/expenditures/editLockedExpenditureMotion.ts:57]: Could all this logic be extracted into a helper considering it's very similar to edit expenditure via permissions?

File: src/redux/sagas/motions/expenditures/editLockedExpenditureMotion.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  meta,
+  meta: { setTxHash, id },
+}: Action<ActionTypes.MOTION_EDIT_LOCKED_EXPENDITURE>) {
+  const colonyManager = yield getColonyManager();
+  const colonyClient = yield colonyManager.getClient(
+    ClientType.ColonyClient,
+    colonyAddress,
+  );
+
+  const batchId = 'motion-edit-locked-expenditure';
```
Comment (jakubcolony) [inline src/redux/sagas/motions/expenditures/editLockedExpenditureMotion.ts:51]: Could you align this saga with the fixes I did in #2020 (e.g. https://github.com/JoinColony/colonyCDapp/pull/2020/files#diff-5097462c62b8afbf61c34f5ba821182563002e86f110568703f0e0b762948b4d) - especially around the batch ID and `createGroupTransaction` params?

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Hey @iamsamgibbs, great work so far!

I finalized the motion but the values in the DB are not 100% correct (claim delay for new slot is null, the recipient address is wrong):
image


I also noticed there's still one requirement that needs addressing:
> The changes proposed via a motion should be stored and made available to the UI.

According to designs (https://www.figma.com/file/5V8pr7iMwXsT9L3VAZsmUt/Colony-v5?type=design&node-id=3615-356361&mode=design&t=iPwrs8mSpxlkFqFa-0), we need to store all changes that the motion is proposing to make.

Feel free to drag me into a call and we can discuss this in more detail.

**Review by jakubcolony — APPROVED** [review]
Really nice work @iamsamgibbs, tested different cases and all seems to be working great 👌 

Before finalizing:
Screenshot 2024-03-11 at 22 37 15

After: 
image

Deleting a slot also works fine, only the modified (deleted) slot is stored for the motion:
image

And the payouts update correctly after finalization:
image

**Review by ArmandoGraterol — APPROVED** [review]

**Review by jakubcolony — APPROVED** [review]
LGTM 👍

**Comment by jakubcolony** [comment]
Are we good to merge this PR @iamsamgibbs?

**Pull request merged**